pkg/pillar/cmd/downloader/syncop.go

```diff
@@ -192,6 +193,10 @@ func handleSyncOp(ctx *downloaderContext, key string,
 			if err != nil {
 				break
 			}
+			// local datastore using IP address in URL, skip the verify if scheme is HTTPS
```

Comment:
This doesn't ,make sense; I thought you would use the specified certificate in this case.

Reply:
Assume we are settled on the local datastore implementation. the URL will be an IP address related string instead of a DNS name (also '.local' name also will not work in a certificate). Even we have the certificate from the local datastore, the https protocol will prevent that. the certificate chain does not work to verify this and will get error like: x509: cannot validate certificate for 192.168.86.76 because it doesn't contain any IP SANs; and although there are some special tricks to make the IP address into the server certificate itself for the local datastore, but that breaks the notion of we don't want to statically define the IP address of a local datastore that was the reason we went to adynamic mechanism with mDNS.

Reply:
But that's no reason to skipVerify (because that doesn;t do any cert checks hence works with anty bogus cert, which is not what we want - we want it to be secure by checking that cert).
In the non-mDNS case things should work without skipVerify, right?
(server name is muble.example.com)

If we want cerificate checks to work with .local it does mean that either
1. we don't and http should be used instead of https
2.  there is a certificate assigned to mumble.local and the cert chain validation checks out from it to the additional certificate specified in the API for the datastore
3. there is a certificate with a subject Altname containing the IP address that we should validate against the additional cert.

For #1 we'd need to just document to use http with .local.
I don't think #3 makes much sense and I have no idea whether the golang TLS code validates using such certificates (in general they can't be trusted since anybody can add a subject altname to their certs)

If #1 is not sufficient and we want #2, that means we need to feed the mDNS IP address into the logic in a similar way that the IP address comes in with DNS lookups i.e., using the golang resolver interface. Thus http/tls would see the server/URL with mumble.local, and use the resolver hook (have your own resolver which does mDNS for .local) so that TCP gets the IP address.

Do you think we need to do #2 or is #1 sufficient for .local?

Reply:
right, non-mDNS case specifying a self-signed cert chain is good to verify.
1) is fine, since it's local in the device, there is no reason to use https.
2) not sure how this will work. the mDNS case, our URL is an ip address.
3) will work, assume users takes care of call the static ip address assignment on the NI for the App, etc.
Thus I can remove this 'skipVerify' option, and let this to be a documentation issue.

Reply:
https://wiki.lfedge.org/display/EVE/Local+DataStore+with+ZeroConfig is updated on this protocol scheme

Reply:
> 2\. not sure how this will work. the mDNS case, our URL is an ip address eventually. and currently even if we install 'avahi' on EVE, there are things missing in alpine glibc so the resolve won't work from my understanding

You have to plug in the mdns resolution in the golang resolver hook. We currently specify a resolver to make sure normal DNS queries go out the correct interface aka use the correct source IP. Presumably that can be used to plug in mDNS as well. Thus the URL as seen by http/tls etc will be https://mumble.local/

> 3\. will work, assume user takes care of all the static ip address assignment on the NI for the App, certificates specifically has the altSANname to this ip-address, etc.

Have you checked that the golang verification of certificate chains handles IP addresses in subjectAltName? I haven't, but this is not a common usage pattern. Thus one would potentially need a patch to golang's core packages for this to work.
Or were you assuming that we'd write new certificate verification code externally to the x509 package, and pass in skipVerify? I don't think that is sufficient (but I could be wrong) since the verification happens as part of the TLS exchange since additional intermediate certificates might be needed in the general case.

>     Thus I can remove this 'skipVerify' option, and let this to be a documentation issue for now.

Makes sense.

Reply:
> Have you checked that the golang verification of certificate chains handles IP addresses in subjectAltName? 

yes I did, here was an example:
 in [ v3_ca ] section, define this ‘subjectAltName = IP:192.168.86.76’, and cli using ‘-extensions v3_ca’,
            X509v3 Basic Constraints: critical
                CA:TRUE
            X509v3 Subject Alternative Name:
                IP Address:192.168.86.76
            X509v3 Key Usage: critical
                Digital Signature, Certificate Sign, CRL Sign
10:22
i’m able to download image using ip address this way (the 'downloader' url map):
 interface: eth0
 Success: 2   Last Success: 2021-06-16T17:17:51.277801611Z
   https://192.168.86.76/images/ubuntu-core-18.qcow2
     Recv KBytes: 543293   sent Kbytes: 1   SentMsg: 1   Total time(sec): 46

Reply:
OK, didn't recall seeing code for this. I assume this was with skipVerify being set.

Reply:
No, that example i was just testing using an external server with url as IP address to verify that in case the local DS needs that, it would work. The code does not need to change from the current, but the user configuration better make sure that eth ip address does not change if using HTTPS.